File: src/release.rb
require 'json'
require 'date'
require 'set'

def read_json(path)
  str = ''
  File.open(path, 'r:utf-8') do |f|
    f.each_line do |line|
      str << line
    end
  end
  JSON.parse(str)
end

def assert(condition, message = '')
  return if condition

  puts message
  exit(1)
end

def check_name(name)
  n = name.gsub(/\s+/, '')
  puts "名前にカタカナ以外が含まれます：#{name}" unless n.match(/^[\p{Katakana}ー]+$/)
  n
end

old_data = read_json('data.json')
old_owner = old_data['owner']
ids = Set.new
max_id = (old_owner['chara'] + old_owner['support']).map do |e|
  assert !ids.member?(e['id']), "id duplicated #{JSON.dump(e)}"
  ids.add(e['id'])
  e['id']
end.max

scenario = ['共通']
old_data['event'].map { |e| e['owner'] }.select { |o| o['type'] == 'scenario' }.each do |o|
  name = o['name']
  unless scenario.include?(name)
    scenario << name
    puts "existing scenario #{name}"
  end
end

owner = read_json('src/icon.json')
owner['chara'].map! do |c|
  name = check_name(c['n'])
  i = c['i']
  list = [i]
  second = "#{File.basename(i, '.png')}_e.png"
  list << second if File.exist?("icon/#{second}")

  idx = old_owner['chara'].index { |e| e['name'] == name }
  id = max_id + 1
  if idx
    id = old_owner['chara'].delete_at(idx)['id']
  else
    puts "new chara detected #{JSON.dump(c)}"
    max_id += 1
  end

  {
    'id' => id,
    'name' => name,
    'icon' => list
  }
end
assert old_owner['chara'].empty?, "chara not found #{JSON.dump(old_owner['chara'])}"

owner['support'].map! do |s|
  name = check_name(s['n'])
  type = s['l'][0..1]
  rarity = s['l'].match(/S{0,2}R/)[0]
  idx = old_owner['support'].index do |e|
    e['name'] == name && e['type'] == type && e['rarity'] == rarity
  end
  id = max_id + 1
  if idx
    id = old_owner['support'].delete_at(idx)['id']
  else
    puts "new support detected #{JSON.dump(s)}"
    max_id += 1
  end

  {
    'id' => id,
    'name' => name,
    'icon' => s['i'],
    'type' => type,
    'rarity' => rarity
  }
end
assert old_owner['support'].empty?, "support not found #{JSON.dump(old_owner['support'])}"

event = read_json('src/event.json').map do |e|
  name = e['n']
  cls = e['c']
  owner_id = nil
  owner_type = nil
  if cls == 'm' || name == '共通'
    # イベント所有者なし
    unless scenario.include?(name)
      puts "new scenario detected #{name}"
      scenario << name
    end
    puts "scenario #{name} (event name:#{e['e']})"
    owner_type = 'scenario'
  elsif cls == 'c'
    # 育成キャラ
    name = check_name(name)
    o = owner['chara'].find { |c| c['name'] == name }
    owner_type = 'chara'
    owner_id = o['id']
  elsif cls == 's'
    # サポートカード
    name = check_name(name)
    type = e['l'][0..1]
    rarity = e['l'].match(/S{0,2}R/)[0]
    o = owner['support'].find do |s|
      s['name'] == name && s['type'] == type && s['rarity'] == rarity
    end
    owner_type = 'support'
    owner_id = o['id']
  else
    assert false, "invalid event type 'e' #{JSON.dump(e)}"
  end
  o = {
    'type' => owner_type,
    'name' => name
  }
  o['id'] = owner_id if owner_id
  {
    'title' => e['e'],
    'title_kana' => e['k'],
    'owner' => o,
    'choices' => e['choices'].map do |c|
      {
        'name' => c['n'],
        'message' => c['t']
      }
    end
  }
end

data = {
  'event' => event,
  'owner' => owner
}

File.open('data.json', 'w') do |file|
  file.write(JSON.pretty_generate(data))
end

size = File.size('data.json')
d = Date.today
v = d.year * 10_000 + d.month * 100 + d.day

info = {
  'version' => v,
  'size' => size
}

File.open('info.json', 'w') do |file|
  file.write(JSON.pretty_generate(info))
end

File: src/release.ts
import * as fs from 'fs';
import * as path from 'path';
import type { Event, IconData, ReleaseOwner, ReleaseData, ReleaseEvent, ReleaseChoice } from './schema.js';
import { ReleaseSupportOwnerSchema, ReleaseDataSchema } from './schema.js';

function readJson<T>(filePath: string): T {
  const content = fs.readFileSync(filePath, 'utf-8');
  return JSON.parse(content);
}

function assert(condition: boolean, message: string = ''): void {
  if (!condition) {
    console.error(message);
    process.exit(1);
  }
}

function checkName(name: string): string {
  // 空白文字が名前に混入して失敗する場合あり
  return name.replace(/\s+/g, '');
}

function main() {
  const oldData = readJson<ReleaseData>('data.json');
  const oldOwner = oldData.owner;
  const ids = new Set<number>();
  let maxId = Math.max(
    ...[...oldOwner.chara, ...oldOwner.support].map((e) => {
      assert(!ids.has(e.id), `id duplicated ${JSON.stringify(e)}`);
      ids.add(e.id);
      return e.id;
    })
  );

  const scenario: string[] = ['共通'];
  oldData.event
    .map((e) => e.owner)
    .filter((o) => o.type === 'scenario')
    .forEach((o) => {
      const name = o.name;
      if (!scenario.includes(name)) {
        scenario.push(name);
        console.log(`existing scenario ${name}`);
      }
    });

  const ownerData = readJson<IconData>('src/icon.json');
  const owner: ReleaseOwner = {
    chara: ownerData.chara.map((c) => {
      const name = checkName(c.n);
      const i = c.i;
      const list: string[] = [i];
      const second = `${path.basename(i, '.png')}_e.png`;
      if (fs.existsSync(`icon/${second}`)) {
        list.push(second);
      }

      // nameの変更を考慮
      const idx = oldOwner.chara.findIndex(
        (e) => e.name === name || JSON.stringify(e.icon) === JSON.stringify(list)
      );
      let id = maxId + 1;
      if (idx !== -1) {
        id = oldOwner.chara.splice(idx, 1)[0].id;
      } else {
        console.log(`new chara detected ${JSON.stringify(c)}`);
        maxId += 1;
      }

      return {
        id,
        name,
        icon: list,
      };
    }),
    support: ownerData.support.map((s) => {
      const name = checkName(s.n);
      const type = s.l.slice(0, 2);
      const rarityMatch = s.l.match(/S{0,2}R/);
      assert(rarityMatch !== null, `Invalid rarity: ${s.l}`);
      const rarity = rarityMatch![0];
      const idx = oldOwner.support.findIndex(
        (e) => e.name === name && e.type === type && e.rarity === rarity
      );
      let id = maxId + 1;
      if (idx !== -1) {
        id = oldOwner.support.splice(idx, 1)[0].id;
      } else {
        console.log(`new support detected ${JSON.stringify(s)}`);
        maxId += 1;
      }
      // zodスキーマでバリデーション
      return ReleaseSupportOwnerSchema.parse({
        id,
        name,
        icon: s.i,
        type,
        rarity,
      });
    }),
  };

  assert(oldOwner.chara.length === 0, `chara not found ${JSON.stringify(oldOwner.chara)}`);
  assert(oldOwner.support.length === 0, `support not found ${JSON.stringify(oldOwner.support)}`);

  const events = readJson<Event[]>('src/event.json');
  const event: ReleaseEvent[] = events.map((e) => {
    const title = e.e;
    const title_kana = e.k;
    const name = checkName(e.n);
    const cls = e.c;
    const choices: ReleaseChoice[] = e.choices.map((c) => ({
      name: c.n,
      message: c.t,
    }));
    if (cls === 'm' || name === '共通') {
      // イベント所有者なし
      if (!scenario.includes(name)) {
        console.log(`new scenario detected ${name}`);
        scenario.push(name);
      }
      console.log(`scenario ${name} (event name:${e.e})`);
      return {
        title,
        title_kana,
        owner: {
          type: 'scenario',
          name,
        },
        choices,
      }
    } else if (cls === 'c') {
      // 育成キャラ
      const o = owner.chara.find((c) => c.name === name);
      assert(o !== undefined, `育成キャラが見つかりません：${name}`);
      return {
        title,
        title_kana,
        owner: {
          type: 'chara',
          name,
          id: o!.id,
        },
        choices,
      }
    } else if (cls === 's') {
      // サポートカード
      const type = e.l?.slice(0, 2) || '';
      const rarityMatch = e.l?.match(/S{0,2}R/);
      assert(rarityMatch !== null, `Invalid rarity: ${e.l}`);
      const rarity = rarityMatch![0];
      const o = owner.support.find((s) => s.name === name && s.type === type && s.rarity === rarity);
      assert(o !== undefined, `サポーターが見つかりません：${name} ${type} ${rarity}`);
      return {
        title,
        title_kana,
        owner: {
          type: 'support',
          name,
          id: o!.id,
        },
        choices,
      }
    } else {
      throw new Error(`invalid event type 'e' ${JSON.stringify(e)}`);
    }
  });

  // zodスキーマでバリデーション
  const validatedData = ReleaseDataSchema.parse({
    event,
    owner,
  });

  fs.writeFileSync('data.json', JSON.stringify(validatedData, null, 2), 'utf-8');

  const size = fs.statSync('data.json').size;
  const d = new Date();
  const v = d.getFullYear() * 10000 + (d.getMonth() + 1) * 100 + d.getDate();

  const info = {
    version: v,
    size,
  };

  fs.writeFileSync('info.json', JSON.stringify(info, null, 2), 'utf-8');
}

main();

